File: CLIENT/LARAVEL/app/Services/GeoLicense/LicenseActivationService.php
<?php

namespace App\Services\GeoLicense;

use Illuminate\Support\Facades\Http;
use Illuminate\Support\Facades\Log;

class LicenseActivationService
{
    public function __construct(
        private readonly LicenseHolder      $holder,
        private readonly MachineIdGenerator $machineId,
    ) {}

    /**
     * Calls POST /api/v1/licenses/activate on the GeoLicense server.
     * Stores the returned token in the cache on success.
     *
     * @throws \RuntimeException if activation fails or the server is unreachable
     */
    public function activate(): void
    {
        $serverUrl  = config('geolicense.server_url');
        $licenseKey = config('geolicense.license_key');

        if (empty($licenseKey)) {
            throw new \RuntimeException('GeoLicense: GEOLICENSE_LICENSE_KEY is not set.');
        }

        $response = Http::timeout(15)
            ->acceptJson()
            ->post("{$serverUrl}/api/v1/licenses/activate", [
                'licenseKey' => $licenseKey,
                'machineId'  => $this->machineId->generate(),
                'osInfo'     => php_uname('s') . ' ' . php_uname('r'),
            ]);

        if ($response->failed() || !$response->json('status')) {
            $message = $response->json('messages.0', 'activation rejected');
            throw new \RuntimeException("GeoLicense activation failed: {$message}");
        }

        $token = $response->json('payload.token');

        if (empty($token)) {
            throw new \RuntimeException('GeoLicense: activation response is missing the token.');
        }

        $this->holder->setToken($token);
        $this->holder->setValid(true);

        Log::info('GeoLicense: license activated successfully.', [
            'key_prefix' => substr($licenseKey, 0, 8) . '****',
        ]);
    }
}

File: CLIENT/SPRINGBOOT PLUGIN/src/main/java/com/alexistdev/geolicense/starter/service/LicenseActivationService.java
package com.alexistdev.geolicense.starter.service;

import com.alexistdev.geolicense.starter.properties.GeoLicenseProperties;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;

public class LicenseActivationService {

    private static final Logger log = LoggerFactory.getLogger(LicenseActivationService.class);

    private final GeoLicenseProperties properties;
    private final LicenseHolder licenseHolder;
    private final MachineIdGenerator machineIdGenerator;
    private final ObjectMapper objectMapper = new ObjectMapper();
    private final HttpClient httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .build();

    public LicenseActivationService(GeoLicenseProperties properties,
                                    LicenseHolder licenseHolder,
                                    MachineIdGenerator machineIdGenerator) {
        this.properties = properties;
        this.licenseHolder = licenseHolder;
        this.machineIdGenerator = machineIdGenerator;
    }

    public void activate() {
        String machineId = machineIdGenerator.generate();
        String osInfo = System.getProperty("os.name") + " " + System.getProperty("os.version");

        String body = objectMapper.createObjectNode()
                .put("licenseKey", properties.getLicenseKey())
                .put("machineId", machineId)
                .put("osInfo", osInfo)
                .toString();

        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(properties.getServerUrl() + "/api/v1/licenses/activate"))
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(body))
                    .timeout(Duration.ofSeconds(15))
                    .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                throw new IllegalStateException("License activation failed — HTTP " + response.statusCode()
                        + ": " + response.body());
            }

            JsonNode root = objectMapper.readTree(response.body());
            boolean status = root.path("status").asBoolean(false);

            if (!status) {
                String message = root.path("messages").path(0).asText("activation rejected");
                throw new IllegalStateException("License activation rejected: " + message);
            }

            String token = root.path("payload").path("token").asText(null);
            if (token == null || token.isBlank()) {
                throw new IllegalStateException("License activation response missing token");
            }

            licenseHolder.setToken(token);
            licenseHolder.setValid(true);
            log.info("License activated successfully for key: {}", maskKey(properties.getLicenseKey()));

        } catch (IllegalStateException e) {
            throw e;
        } catch (Exception e) {
            throw new IllegalStateException("Failed to reach license server at " + properties.getServerUrl(), e);
        }
    }

    private String maskKey(String key) {
        if (key == null || key.length() < 8) return "***";
        return key.substring(0, 8) + "****";
    }
}
